---
title: "EDA_primer_instructor"
date: "02/27/2019"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

```{r, echo=FALSE, out.width="75%", out.height="75%", fig.align='center'}
knitr::include_graphics("Monet Logo-12-06-24_cropped_dpi600.png")
```

\pagebreak

> **Learning Outcomes**
> 
> * Use RStudio and R Notebooks to write & execute R code
> * Manipulate data and make multiple types of plots


-----

> 1. The R ecosystem: R | RStudio | R Notebooks | R Packages
> 2. File/data input/output
> 3. Data cleaning

## Some useful functions in this primer

### Reading & writing text files with readr
`read_csv`, `write_csv`
`read_tsv`, `write_tsv`
`read_delim`, `write_delim`

### Wrangling data with dplyr
`filter`, `select`, `distinct`, `arrange`, `rename`, `mutate`, `transmute`

### Tidying data with tidyr
`gather`, `spread`, `separate`, `unite`


```{r}
# install and/or load tidyverse
#install.packages("tidyverse")
library(tidyverse)
```

## Read and look at some data
What are the columns, how many rows are there, how many columns?

```{r}
url <- "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE69nnn/GSE69360/suppl/GSE69360_RNAseq.counts.txt.gz"
gse69360 <- read_tsv(url)

head(gse69360, 20)
colnames(gse69360)
nrow(gse69360)
ncol(gse69360)
```

## An example of some exploratory data analysis using filter and select

```{r}
# keep only chromosome 1
gse69360 %>%
    filter(grepl("chr1", Chr))
# keep genes that have a length <= 50
gse69360 %>%
    filter(Length <= 50)
# keep genes that satisfy both of the conditions above
gse69360 %>%
    filter(Length <= 50 & grepl("chr1", Chr))
# keep columns Geneid and any column the has the characters "F_"
gse69360 %>%
    select(Geneid, matches("F_"))
# drop columns, Chr, Start, End, Strand, Length
gse69360 %>%
    select(-Chr, -Start, -End, -Strand, -Length)
# drop all columns that have a letter A thru F, followed by an underscore
gse69360 %>%
    select(-matches("[AF]_"))
```
Exercise

```{r}
# keep Geneid, Chr, Length and all column that start have first letter of B thru F and
# a second letter of F followed by an underscore
# keep only rows with chr1 in Chr OR have a length less than or equal to 100

gsenew <- gse69360 %>%
    select(Geneid, Chr, Length, matches("[BF]F_")) %>%
    filter(grepl("chr1", Chr) | Length <= 100)
gsenew
```

## EDA with mutate, distinct arrange

```{r}

# with mutate can overwrite column by giving the same column name
# could also use transmutate for the same purpose

# get rid of info in gene from period to the end and make a new column called Geneid2
gse69360 %>%
    mutate(Geneid2 = gsub("\\.[0-9]*$", "", Geneid))
# get rid of info in gene from period to the end and overwrite old column
gse69360 %>%
    mutate(Geneid = gsub("\\.[0-9]*$", "", Geneid))
# find the rows with unique Chr
gse69360 %>%
    distinct(Chr)
# find rows with unique Strand
gse69360 %>%
    distinct(Strand)
# find rows with unique Chr and Stands
gse69360 %>%
    distinct(Chr, Strand)
# arrange data frame based on Chr
gse69360 %>%
    arrange(desc(Chr))
# arrange data frame based on Chr and then Length
gse69360 %>%
    arrange(Chr, Length)

```

### Exercise

helpful functions; `rowSums` and `colnames`

```{r}
# Make a new column that is the total expression for each gene across all samples per sample
# drop all columns except Geneid and total expression column
# Then sort the table based on the total expression, with highest expressed genes on top

# ADD CODE HERE
```

### Tidying data with tidyr

Dataset details:

- AA: Agilent Adult; AF: Agilent Fetus
- BA: BioChain Adult; BF: BioChain Fetus
- OA: OriGene Adult
- Tissues: Colon, Heart, Kidney, Liver, Lung, Stomach

```{r}
# Tidy the table to have a sample column and a column for the counts
gse69360 %>%
    select(Geneid, matches("[AF]_")) %>%
    gather(-Geneid, key="Sample", value="Counts")

# Tidy and then undo the tidying
gse69360 %>%
    select(Geneid, matches("[AF]_")) %>%
    gather(-Geneid, key="Sample", value="Counts") %>%
    spread(key = "Sample", value = "Counts")

# Tidy and then separate the sample column
gse69360 %>%
    select(Geneid, matches("[AF]_")) %>%
    gather(-Geneid, key="Sample", value="Counts") %>%
    separate(Sample, into=c("Source_Stage", "Tissue"), sep="_")

# Separate the Source stage column int Source and Stage
gse69360 %>%
    select(Geneid, matches("[AF]_")) %>%
    gather(-Geneid, key="Sample", value="Counts") %>%
    separate(Sample, into=c("Source_Stage", "Tissue"), sep="_") %>%
    separate(Source_Stage, into=c("Source", "Stage"), sep=1)

# Make new column that has the Stage and tissue united
# Defualt is underscore
gse69360 %>%
    select(Geneid, matches("[AF]_")) %>%
    gather(-Geneid, key="Sample", value="Counts") %>%
    separate(Sample, into=c("Source_Stage", "Tissue"), sep="_") %>%
    separate(Source_Stage, into=c("Source", "Stage"), sep=1) %>%
    unite(Stage_Tissue, Stage, Tissue)
```

### Exercise

What we want to do is to get the data ready for plotting. This means that we want to do the following to the data

1. Get the logcpm of the counts (won't be ever using the Chr, Start, End, Strand, Length columns from here on out so just drop them)
2. Put it in tidy form (i.e. gather into sample and counts columns, we won't care about anything but the samples and geneid so drop the rest)
3. Split sample into source, stage, and tissue columns

```{r}

# ADD CODE HERE

```

# Plotting the distribution of gene-expressions

In the first plot when can plot a scatter plot comparing the expression between to samples

```{r}
gene_logcpm %>%
    filter(Sample == "AF_Colon" | Sample == "BF_Colon") %>%
    select(Geneid, Sample, Logcpm) %>%
    spread(key = Sample, value = Logcpm) %>%
    ggplot(aes(x = AF_Colon, y = BF_Colon)) +
    geom_point(alpha = 0.1, size = 0.5) +
    geom_smooth(method=lm) +
    theme_minimal()
```

### Exercise

Make a boxplot plot of the expression values for every sample. Differenitate in the plot between both
the different types of tissues and the different stages

```{r}
gene_logcpm %>%
    ggplot(aes(x = Sample, y = Logcpm, color = Tissue, linetype = Stage)) +
    geom_boxplot(outlier.size = 0.2, outlier.shape = 0.2) +
    scale_y_continuous(limits = c(0, 1)) +
    coord_flip() +
    theme_minimal()
```

For the next plots we will need to look at the genes locations. The new dataframe
is built with the code below

```{r}
# look at orignal data
head(gse69360)

# wrangle the data
gene_loc <- gse69360 %>%
    select(-matches("[AF]_")) %>%
    mutate(Geneid = gsub("\\.[0-9]*$", "", Geneid)) %>%
    mutate(Chr = gsub(";.*$", "", gse69360$Chr)) %>%   # a gene should all be on the same chromosme
    mutate(Start = as.numeric(gsub(";.*$", "", gse69360$Start))) %>% # I think these are the different entrons, just look at the first
    mutate(End = as.numeric(gsub(";.*$", "", gse69360$End))) %>%
    mutate(Strand = gsub(";.*$", "", gse69360$Strand)) # I'm not sure what Strand means

# Change the Chr column into a factor
# This gives the ability to set relational orders between character data. Useful when ordering plot variables
# or making comparisons between character data types
gene_loc$Chr <- factor(gene_loc$Chr,
                       levels = paste("chr",
                                      c((1:22), "X", "Y", "M"),
                                      sep=""))

# look at new data
head(gene_loc,n=20)
```

Make a bar plot of the number of genes in each chromosome

```{r}
gene_loc %>%
    ggplot(aes(x = Chr)) +
    geom_bar() +
    coord_flip() +
    theme_minimal() +
    labs(title = "No. genes per chromosome",
         x = "Chromosome",
         y = "No. of genes")
```


## Excerise

Make a histogram plot hwere there is a histogram for each chromosome
that shows the number of genes as a function of the gene length

**Help** May want to use facet_wrap and put the x-axis in log10 scale

```{r}

# ADD Code here

```


The last thing we will look at is plotting gene with high variance in expression across the samples.
The first thing to do is to determine which gene have high variance

```{r}
# Find the total number of genes
num_totgenes <- gene_logcpm %>%
    distinct(Geneid) %>%
    nrow()

# Find the high-var genes
highvar_genes <- gene_logcpm %>%
    group_by(Geneid) %>%
    summarize(iqr = IQR(Logcpm)) %>%
    top_n((ceiling(num_totgenes*0.05)), iqr) %>%
    pull(Geneid)

# Show how many genes there are
length(highvar_genes)

```

# Plotting expression of high-var chr1 genes across samples

This will shower more readily which samples are different from each other

```{r}
chry_highvar_genes <- gene_loc %>%
    filter(Chr == "chr1" & Geneid %in% highvar_genes) %>%
    pull(Geneid)

plot_chry_highvar_boxplot <- gene_logcpm %>%
    filter(Geneid %in% chry_highvar_genes) %>%
    ggplot(aes(x = reorder(Sample, Logcpm, FUN = median),
               y = Logcpm,
               color = Sample)) +
    geom_boxplot() +
    coord_flip() +
    theme_minimal() +
    theme(legend.position = "none")
plot_chry_highvar_boxplot
```
